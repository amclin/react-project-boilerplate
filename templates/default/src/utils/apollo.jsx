/* eslint-disable react/jsx-props-no-spreading, no-console */
/* FOR SERVER SIDE RENDERING APOLLO FETCHES: Use unmodifed link below.
 * https://github.com/zeit/next.js/blob/canary/examples/with-apollo/lib/apollo.js
 * this file is modified from link above. Had some issues getting it to work with hooks.
 */
import React from 'react'
import { ApolloProvider } from '@apollo/react-hooks'
import { ApolloClient } from 'apollo-client'
import { InMemoryCache } from 'apollo-cache-inmemory'
import { HttpLink } from 'apollo-link-http'
import fetch from 'isomorphic-unfetch'
import getConfig from 'next/config'

const { publicRuntimeConfig } = getConfig()
const { apiGatewayUrl } = publicRuntimeConfig
let apolloClient = null

/**
 * Creates and configures the ApolloClient
 * @param  {Object} [initialState={}]
 */
function createApolloClient(initialState = {}) {
  return new ApolloClient({
    ssrMode: false, // !process.browser
    link: new HttpLink({
      uri: apiGatewayUrl, // Server URL (must be absolute)
      // credentials: 'same-origin', // Additional fetch() options like `credentials` or `headers`
      fetch
    }),
    cache: new InMemoryCache().restore(initialState)
  })
}

/**
 * Always creates a new apollo client on the server
 * Creates or reuses apollo client in the browser.
 * @param  {Object} initialState
 */
function initApolloClient(initialState) {
  // Make sure to create a new client for every server-side request so that data
  // isn't shared between connections (which would be bad)

  if (!process.browser) {
    return createApolloClient(initialState)
  }

  // Reuse client on the client-side
  if (!apolloClient) {
    apolloClient = createApolloClient(initialState)
  }

  return apolloClient
}

/**
 * Creates and provides the apolloContext
 * to a next.js PageTree. Use it by wrapping
 * your PageComponent via HOC pattern.
 * @param {Function|Class} PageComponent
 * @param {Object} [config]
 */
export function withApollo(PageComponent) {
  // eslint-disable-next-line
  const WithApollo = ({ apolloClient, apolloState, ...pageProps }) => {
    const client = apolloClient || initApolloClient(apolloState)
    return (
      <ApolloProvider client={client}>
        <PageComponent {...pageProps} />
      </ApolloProvider>
    )
  }

  if (PageComponent.getInitialProps) {
    WithApollo.getInitialProps = async ctx => {
      // Initialize ApolloClient, add it to the ctx object so
      // we can use it in `PageComponent.getInitialProp`.
      // eslint-disable-next-line
      const apolloClient = (ctx.apolloClient = initApolloClient())

      // Run wrapped getInitialProps methods
      const pageProps = await PageComponent.getInitialProps(ctx)

      // Only on the server:
      if (!process.browser) {
        // When redirecting, the response is finished.
        // No point in continuing to render
        if (ctx.res && ctx.res.finished) {
          return pageProps
        }
      }

      // Extract query data from the Apollo store
      const apolloState = apolloClient.cache.extract()

      return {
        ...pageProps,
        apolloState
      }
    }
  }

  return WithApollo
}

export default withApollo
