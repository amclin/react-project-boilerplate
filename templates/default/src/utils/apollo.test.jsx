import React, { Component } from 'react'
import { render } from '@testing-library/react'
import { withApollo } from './apollo'

jest.mock('next/config', () => () => ({
  publicRuntimeConfig: {
    apiGatewayUrl: process.env.API_GATEWAY_URL
  }
}))

// eslint-disable-next-line
class AppWithProps extends Component {
  render() {
    return <div data-testid="sample-app">OK!</div>
  }
}

// eslint-disable-next-line
const AppWithNoProps = () => <div data-testid="sample-app">OK!</div>

describe('apollo', () => {
  it('server side', () => {
    const SampleComponentWithApollo = withApollo(AppWithNoProps)
    const { asFragment, getByTestId } = render(<SampleComponentWithApollo />)

    expect(getByTestId('sample-app')).toBeTruthy()
    expect(asFragment()).toMatchSnapshot()
  })

  it('server side with getInitialProps unfinished', () => {
    const mockCtx = {}
    AppWithProps.getInitialProps = jest.fn().mockResolvedValue(mockCtx)

    const SampleComponentWithApollo = withApollo(AppWithProps)
    const { asFragment, getByTestId } = render(<SampleComponentWithApollo />)

    expect(SampleComponentWithApollo.getInitialProps(mockCtx)).toMatchSnapshot()
    expect(getByTestId('sample-app')).toBeTruthy()
    expect(asFragment()).toMatchSnapshot()
  })

  it('server side with getInitialProps finished', () => {
    const mockCtx = { res: { finished: true } }
    AppWithProps.getInitialProps = jest.fn().mockResolvedValue(mockCtx)

    const SampleComponentWithApollo = withApollo(AppWithProps)
    const { asFragment, getByTestId } = render(<SampleComponentWithApollo />)

    expect(SampleComponentWithApollo.getInitialProps(mockCtx)).toMatchSnapshot()
    expect(getByTestId('sample-app')).toBeTruthy()
    expect(asFragment()).toMatchSnapshot()
  })

  it('client side with getInitialProps', () => {
    jest.resetModules() // this is important - it clears the cache
    process.browser = true

    const mockCtx = {}
    AppWithProps.getInitialProps = jest.fn().mockResolvedValue(mockCtx)

    const SampleComponentWithApollo = withApollo(AppWithProps)
    const { asFragment, getByTestId } = render(<SampleComponentWithApollo />)

    expect(SampleComponentWithApollo.getInitialProps(mockCtx)).toMatchSnapshot()
    expect(getByTestId('sample-app')).toBeTruthy()
    expect(asFragment()).toMatchSnapshot()

    jest.resetModules()
  })
})
