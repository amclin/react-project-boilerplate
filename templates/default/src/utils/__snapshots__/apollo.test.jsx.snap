// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apollo client side with getInitialProps 1`] = `Promise {}`;

exports[`apollo client side with getInitialProps 2`] = `
<DocumentFragment>
  <div
    data-testid="sample-app"
  >
    OK!
  </div>
</DocumentFragment>
`;

exports[`apollo server side 1`] = `
<DocumentFragment>
  <div
    data-testid="sample-app"
  >
    OK!
  </div>
</DocumentFragment>
`;

exports[`apollo server side with getInitialProps finished 1`] = `Promise {}`;

exports[`apollo server side with getInitialProps finished 2`] = `
<DocumentFragment>
  <div
    data-testid="sample-app"
  >
    OK!
  </div>
</DocumentFragment>
`;

exports[`apollo server side with getInitialProps unfinished 1`] = `Promise {}`;

exports[`apollo server side with getInitialProps unfinished 2`] = `
<DocumentFragment>
  <div
    data-testid="sample-app"
  >
    OK!
  </div>
</DocumentFragment>
`;
